### ISSUE TITLE
IPv4 and IPv6 (dual-stack) support?

### ISSUE THREAD

**Issue by aofei** [GoogleCloudPlatform/compute-archlinux-image-builder#45]
I created a custom VPC network with IPv6 support today and configured my old instance to use it.

Although `ip a s eth0` shows that the interface has an external IPv6 address, I cannot reach any public IPv6 addresses. For example, `ping -6 dns.google` prints `ping: connect: Network is unreachable`.

Then I created a new instance using the `ubuntu-2204-jammy-v20220528` image and everything worked fine. This makes me think that my old instance (created about two years ago) might have some unknown problem. So I used this project to create a new instance:

```bash
$ gcloud compute instances create archlinux-ipv6-test --image-project arch-linux-gce --image-family arch --network-interface subnet=custom,stack-type=IPV4_IPV6
```

Everything worked fine until I rebooted the instance with `sudo reboot`. And after reboot `ip a s eth0` shows no external IPv6 address.

I don't know if I'm doing something wrong, are there any other steps I need to do? Or does this project not support IPv6 yet?

**Comment by aofei** [comment]
Routing problem. I worked around this problem by adding the following line to the `startup-script`:

```bash
ip -6 route add default via $(curl http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/gateway-ipv6 -H "Metadata-Flavor: Google") dev eth0
```

But I still think it should be handled automatically by some system component, so I leave this issue open for now.

**Comment by toastwaffle** [comment]
Hi @aofei.

I'm not familiar with custom VPCs - could you share either gcloud commands or step by step instruction for the cloud console so I can replicate your setup for testing?

**Comment by aofei** [comment]
```bash
$ gcloud compute networks create archimg-issue-45 --subnet-mode custom
$ gcloud compute networks subnets create archimg-issue-45 --network archimg-issue-45 --region asia-east2 --range 10.10.0.0/24 --stack-type IPV4_IPV6 --ipv6-access-type EXTERNAL
$ gcloud compute firewall-rules create archimg-issue-45-allow-ssh --network archimg-issue-45 --allow tcp:22
$ gcloud compute instances create archimg-issue-45 --zone asia-east2-a --image-project arch-linux-gce --image-family arch --network-interface subnet=archimg-issue-45,stack-type=IPV4_IPV6 --metadata block-project-ssh-keys=true --no-shielded-secure-boot --no-shielded-vtpm --no-shielded-integrity-monitoring
```

If all goes well, these four lines of commands should create a problematic instance for you.

Then you can execute:

```bash
$ gcloud compute ssh archimg-issue-45 --zone asia-east2-a --command "ip a s eth0; ping -6 -c 3 dns.google"
```

and everything should be fine at this point. But if you then execute:

```bash
$ gcloud compute ssh archimg-issue-45 --zone asia-east2-a --command "sudo reboot"
```

the problem arises at this point. Either the instance is not reachable at all (no internet), or `ip a s eth0` shows no external IPv6 address. The problem I'm having at this moment is the first one (no internet). The problem I had three days ago was the second.

**Comment by aofei** [comment]
> the problem arises at this point. Either the instance is not reachable at all (no internet), or `ip a s eth0` shows no external IPv6 address.

For the second case (`ip a s eth0` shows no external IPv6 address), a workaround is to add the following line to the `startup-script`:

```bash
ip -6 addr add $(curl -fsS http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6 -H "Metadata-Flavor: Google") dev eth0
```

**Comment by aofei** [comment]
I happened to have the same problem today (`ip a s eth0` shows no external IPv6 address after `sudo reboot`) with an instance I created with `ubuntu-2204-jammy-v20220528` image.

So I don't know, according to the [IPv6 address assignment](https://cloud.google.com/compute/docs/ip-addresses/configure-ipv6-address#ipv6-assignment), this problem seems more likely to be caused by some metadata server bug in `asia-east2` region than the VM OS now. Because I just tested the commands in https://github.com/GoogleCloudPlatform/compute-archlinux-image-builder/issues/45#issuecomment-1147550758 with `us-central1` region, and everything works fine.

If it's a metadata server bug, then this issue can be closed. @toastwaffle
